lib/rom/support/data_proxy.rb

```diff
@@ -0,0 +1,51 @@
+module ROM
+  module DataProxy
+    attr_reader :data, :header, :tuple_proc
+
+    NON_FORWARDABLE = [
+      :each, :to_a, :to_ary, :kind_of?, :instance_of?, :is_a?
+    ].freeze
+
+    def self.included(klass)
+      klass.class_eval do
+        extend ClassMethods
+        include Equalizer.new(:data)
+      end
+    end
+
+    def initialize(data, header, tuple_proc = self.class.tuple_proc)
```

Comment:
Yes, I "blame" Sequel too (I borrowed "dataset" from it). In fact, those may *not be tuples* (like in CSV adapter) so we should come up with a better term here. I guess "row" would be OK since it's "a row in a dataset". This may or may not sound good in a particular adapter's case. For instance "row" works great for CSV obviously.